### ISSUE TITLE
Rollup 0.59.x removes important code from ReactDOM bundle

### ISSUE THREAD

**Issue by gaearon** [rollup/rollup#2199]
Reproducing case:

1. Clone https://github.com/isidrok/react-bug
1. `npm install` 
1. `npm start`
1. Open `build.dev.js` and scroll to `diffProperties$1` function definition
1. Open `node_modules/react-dom/cjs/react-dom.development.js` for comparison and scroll to `diffProperties$1` function definition

You’ll see that the original function definition contains a `for` loop like

```js
function diffProperties$1(domElement, tag, lastRawProps, nextRawProps, rootContainerElement) {
  // ...
  for (propKey in lastProps) {
    if (nextProps.hasOwnProperty(propKey) || !lastProps.hasOwnProperty(propKey) || lastProps[propKey] == null) {
      continue;
    }
    if (propKey === STYLE) {
      // ...
    } else if (propKey === DANGEROUSLY_SET_INNER_HTML || propKey === CHILDREN) {
      // ...
    } else if (registrationNameModules.hasOwnProperty(propKey)) {
      // ...
    } else {
      // ...
    }
    // ...
}
```

However the `diffProperties$1` definition in the Rollup’d `bundle.dev.js` is completely missing the (essential) logic under those conditions as well as conditions themselves, except for the last one. This produces code that behaves differently. Its `for` loop only contains the `registrationNameModules.hasOwnProperty(propKey)` check and not the preceding ones.

I verified that downgrading to 0.58.x fixes it.

**Comment by isidrok** [comment]
Didn't think on downgrading rollup since it was my first time trying out React 16.3... :sweat_smile:

Thank you @gaearon

**Comment by lukastaegert** [comment]
Thanks a lot for posting this issue + the repro. It took me a while to figure out but it seems this is actually a rather old issue that has come to the surface with the enhanced conditional tracking. Basically it appears that `for..in` statements will not count as assignments to their loop variable.

This is a minimal repro which is completely removed:
```javascript
let x = false;

for (x in {key: true}) {
	if (x) console.log('removed');
}
```

I think it should not be too difficult to fix, will take a look soon.

**Comment by lukastaegert** [comment]
Fixed in Rollup v0.59.2, please check it out!

**Comment by isidrok** [comment]
Hi, the reproduction demo works fine with last rollup version, thanks!
